File: javascript/src/alphabet.js
/**
 * XID 文本层：自定义进制编解码。
 */

export class RadixCodec {
  /** @param {string} alphabet */
  constructor(alphabet) {
    const chars = [...alphabet];
    if (chars.length < 2) throw new Error('alphabet must have at least 2 unique characters');
    this.base = chars.length;
    this.chars = chars;
    /** @type {Map<string, number>} */
    this.fromCustom = new Map();
    for (let i = 0; i < chars.length; i++) {
      if (this.fromCustom.has(chars[i])) {
        throw new Error(`alphabet contains duplicate character ${chars[i]}`);
      }
      this.fromCustom.set(chars[i], i);
    }
  }

  /** @param {Uint8Array} data */
  encodeBytes(data) {
    if (data.length === 0) return this.chars[0];
    const wrapped = new Uint8Array(2 + data.length);
    wrapped[0] = (data.length >> 8) & 0xff;
    wrapped[1] = data.length & 0xff;
    wrapped.set(data, 2);
    let n = 0n;
    for (const b of wrapped) n = (n << 8n) | BigInt(b);
    return this.intToString(n);
  }

  /** @param {string} s @returns {Uint8Array} */
  decodeBytes(s) {
    if (!s) throw new Error('empty string');
    const n = this.stringToInt(s);
    let raw = n.toString(16);
    if (raw.length % 2) raw = '0' + raw;
    const bytes = new Uint8Array(raw.length / 2);
    for (let i = 0; i < bytes.length; i++) {
      bytes[i] = parseInt(raw.slice(i * 2, i * 2 + 2), 16);
    }
    for (const pad of [0, 1]) {
      const buf = new Uint8Array(pad + bytes.length);
      buf.set(bytes, pad);
      if (buf.length < 2) continue;
      const dataLen = (buf[0] << 8) | buf[1];
      if (buf.length !== 2 + dataLen) continue;
      return buf.slice(2);
    }
    throw new Error('invalid encoded data length');
  }

  /** @param {bigint} n */
  intToString(n) {
    if (n === 0n) return this.chars[0];
    const base = BigInt(this.base);
    const chars = [];
    while (n > 0n) {
      const rem = Number(n % base);
      chars.push(this.chars[rem]);
      n /= base;
    }
    return chars.reverse().join('');
  }

  /** @param {string} s */
  stringToInt(s) {
    let n = 0n;
    const base = BigInt(this.base);
    for (const ch of s) {
      if (!this.fromCustom.has(ch)) throw new Error(`invalid character ${ch}`);
      n = n * base + BigInt(this.fromCustom.get(ch));
    }
    return n;
  }
}

File: rust/lib/src/alphabet.rs
use std::collections::HashMap;

use num_bigint::BigUint;
use num_traits::{ToPrimitive, Zero};

use crate::error::IdMixError;

pub struct RadixCodec {
    pub base: usize,
    chars: Vec<char>,
    from_custom: HashMap<char, usize>,
}

impl RadixCodec {
    pub fn new(alphabet: &str) -> Result<Self, IdMixError> {
        let chars: Vec<char> = alphabet.chars().collect();
        if chars.len() < 2 {
            return Err(IdMixError::msg(
                "alphabet must have at least 2 unique characters",
            ));
        }
        let mut from_custom = HashMap::with_capacity(chars.len());
        for (i, &ch) in chars.iter().enumerate() {
            if from_custom.insert(ch, i).is_some() {
                return Err(IdMixError::msg(format!(
                    "alphabet contains duplicate character {ch:?}"
                )));
            }
        }
        Ok(Self {
            base: chars.len(),
            chars,
            from_custom,
        })
    }

    pub fn encode_bytes(&self, data: &[u8]) -> Result<String, IdMixError> {
        if data.is_empty() {
            return Ok(self.chars[0].to_string());
        }
        let mut wrapped = Vec::with_capacity(2 + data.len());
        let len = data.len() as u16;
        wrapped.push((len >> 8) as u8);
        wrapped.push((len & 0xff) as u8);
        wrapped.extend_from_slice(data);
        let n = BigUint::from_bytes_be(&wrapped);
        Ok(self.int_to_string(&n))
    }

    pub fn decode_bytes(&self, s: &str) -> Result<Vec<u8>, IdMixError> {
        if s.is_empty() {
            return Err(IdMixError::msg("empty string"));
        }
        let n = self.string_to_int(s)?;
        let raw = n.to_bytes_be();
        for pad in 0..=1 {
            let mut buf = vec![0u8; pad + raw.len()];
            buf[pad..].copy_from_slice(&raw);
            if buf.len() < 2 {
                continue;
            }
            let data_len = u16::from_be_bytes([buf[0], buf[1]]) as usize;
            if buf.len() != 2 + data_len {
                continue;
            }
            return Ok(buf[2..].to_vec());
        }
        Err(IdMixError::msg("invalid encoded data length"))
    }

    fn int_to_string(&self, n: &BigUint) -> String {
        if n.is_zero() {
            return self.chars[0].to_string();
        }
        let base = BigUint::from(self.base as u64);
        let zero = BigUint::from(0u8);
        let mut num = n.clone();
        let mut chars = Vec::new();
        while num > zero {
            let rem = &num % &base;
            num /= &base;
            let idx = rem.to_u32().unwrap_or(0) as usize;
            chars.push(self.chars[idx]);
        }
        chars.iter().rev().collect()
    }

    fn string_to_int(&self, s: &str) -> Result<BigUint, IdMixError> {
        let base = BigUint::from(self.base as u64);
        let mut n = BigUint::from(0u8);
        for ch in s.chars() {
            let idx = *self
                .from_custom
                .get(&ch)
                .ok_or_else(|| IdMixError::msg(format!("invalid character {ch:?}")))?;
            n *= &base;
            n += BigUint::from(idx as u64);
        }
        Ok(n)
    }
}
